lightning-invoice/src/payment.rs

```diff
@@ -552,6 +592,20 @@ where
 			} => {
 				if let Some(short_channel_id) = short_channel_id {
 					let path = path.iter().collect::<Vec<_>>();
+					if let Some(payment_id) = payment_id {
+						// When the failed payment was a probe, we make sure to not penalize the last
+						// hop and then drop the event instead of handing it up to the user's event
+						// handler.
+						if self.payer.payment_is_probe(*payment_hash, *payment_id) {
+							if *rejected_by_dest {
+								self.scorer.lock().probe_successful(&path);
+							} else {
+								self.scorer.lock().probe_failed(&path, *short_channel_id);
+							}
+							return;
```

Comment:
Hm, besides adding some logging (done @ https://github.com/lightningdevkit/rust-lightning/pull/1567/commits/319ad1bb75fdc6c55e5223173e0e88a92b2fd8f8), I can't think of a good example where we'd like to let this run through, i.e., let the probe be handled as a regular payment?

Reply:
I guess, since presumably the user called `send_probe_along_path`, they may like to know what the result is and handle it programmatically (e.g., updating some database if they want to later use the probe for an actual payment, or how, or @TonyGiorgio's use case). Similarly, how they will get a `PaymentSent` or `PaymentFailed` after calling `pay_invoice`, et al.

But by swallowing the event, they can only manually look for logging. If we don't swallow it, then they would need to examine `PaymentPathFailed` event  an interpret (a) whether it is a probe and (b) whether or not it failed. Dedicated events would align the API with `pay_invoice`.

Reply:
Ah, that makes sense. So I'll switch to `ProbeSuccessful`/`ProbeFailed` events and let the events be handed up after training the `Scorer`.